### PR TITLE
Support python 3.9 on Spark 3.1

### PR DESCRIPTION
**Pull request by guoqiao1992**
*Issue #, if available:*
Spark container needs to support python 3.9

*Description of changes:*
* Changed the structure of the directories so that all python 3.9 related changes are independent of python 3.7. More Specifically
  * Created `./smsparkbuild/py{python-version}` (e.g, ./smsparkbuild/py39) that includes 3 files: Pipfile, pyproject.toml and setup.py 
  * Change the path of Dockerfile from `./spark/processing/{spark-version}/py3/docker` to `./spark/processing/{spark-version}/py3/docker/py{python-version}`
* Change the Dockerfile.cpu under `./spark/processing/3.1/py3/docker/py39` to install python 3.9
* Update Makefile to build smspark and the container
* Created a integration test to test python 3.9
* Automatic refactor done by black

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -44,7 +46,7 @@ build-container-library: init
 	cp -- dist/*.whl ${BUILD_CONTEXT}
 
 install-container-library: init
-	pipenv run safety check  # https://github.com/pyupio/safety
+	pipenv run safety check -i 43975 # https://github.com/pyupio/safety
```
Comment (qiyhe) [inline Makefile:49]: Why are we adding this `-i 43975` ?

File: spark/processing/3.1/Pipfile.lock [hunk 2]
```diff
@@ -317,6 +328,13 @@
             "index": "pypi",
             "version": "==3.7.3"
         },
+        "iniconfig": {
+            "hashes": [
+                "sha256:011e24c64b7f47f6ebd835bb12a743f2fbe9a26d4cecaa7f53bc4f35ee9da8b3",
+                "sha256:bc3af051d7d14b2ee5ef9969666def0cd1a000e121eaea580d4a313df4b37f32"
+            ],
```
Comment (qiyhe) [inline spark/processing/3.1/Pipfile.lock:335]: We are exposing some hash functions here, will that be a security risk?

File: smsparkbuild/py39/Pipfile [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+verify_ssl = true
+
+[dev-packages]
+
+[packages]
+tenacity = "==5.1.4"
+psutil = "==5.7.0"
+click = "==8.0.0"
+watchdog = "==0.10.3"
+waitress = "==2.1.1"
```
Comment (qiyhe) [inline smsparkbuild/py39/Pipfile:13]: From online discussion: check and update external dependencies (ex, update boto3 to latest version)

File: spark/processing/3.1/Pipfile [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+[dev-packages]
+
+[packages]
+tenacity = "==5.1.4"
+psutil = "==5.7.0"
+click = "==8.0.0"
+watchdog = "==0.10.3"
+waitress = "==2.1.1"
+requests = "==2.24.0"
+rsa = "==4.3"
```
Comment (qiyhe) [inline spark/processing/3.1/Pipfile:15]: From online discussion: probably need to remove these as they are not being used

File: test/resources/code/scala/hello-scala-spark/hello-scala-spark.sbt [hunk 5]
```diff
@@ -1,9 +1,9 @@
 name := "hello-scala-spark"
 version := "1.0"
-scalaVersion := "2.11.12"
+scalaVersion := "2.12.12"
 useCoursier := false
 retrieveManaged := true
-libraryDependencies += "org.apache.spark" %% "spark-sql" % "2.4.4"
+libraryDependencies += "org.apache.spark" %% "spark-sql" % "3.0.0"
```
Comment (qiyhe) [inline test/resources/code/scala/hello-scala-spark/hello-scala-spark.sbt:6]: should this be 3.0 or 3.1 ? As I saw our latest release will be using spark 3.1

File: spark/processing/3.1/setup.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
+# Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (qiyhe) [inline spark/processing/3.1/setup.py:1]: This file seems duplicated with py39, and py37 is somehow a binary file not visible. 

Are py37 and py39 generated by build? Should we exclude them?

File: smsparkbuild/py39/pyproject.toml [hunk 7]
```diff
@@ -1,3 +1,3 @@
 [tool.black]
 line-length = 120
-target-version = ['py39']
+target-version = ['py37']
```
Comment (qiyhe) [inline smsparkbuild/py39/pyproject.toml:3]: why are we changing target version to37 for python 3.9?

File: Makefile (hunk 1 above)
Comment (guoqiao1992) [reply Makefile:50]: it's an issue for urllib3, bypass for now due to some circular dependency after bump up the version

File: spark/processing/3.1/Pipfile.lock (hunk 2 above)
Comment (guoqiao1992) [reply spark/processing/3.1/Pipfile.lock]: removed

File: smsparkbuild/py39/Pipfile (hunk 3 above)
Comment (guoqiao1992) [reply smsparkbuild/py39/Pipfile:13]: yep updated.

File: spark/processing/3.1/Pipfile (hunk 4 above)
Comment (guoqiao1992) [reply spark/processing/3.1/Pipfile]: they are used in smspark

File: spark/processing/3.1/setup.py (hunk 6 above)
Comment (guoqiao1992) [reply spark/processing/3.1/setup.py]: removed

File: smsparkbuild/py39/pyproject.toml (hunk 7 above)
Comment (guoqiao1992) [reply smsparkbuild/py39/pyproject.toml:3]: changed by mistake. Changed back.

File: new_images.yml [hunk 8]
```diff
@@ -1,7 +1,7 @@
 ---
 new_images:
-  - spark: "3.0.0"
+  - spark: "3.1.1"
     use-case: "processing"
     processors: ["cpu"]
-    python: ["py37"]
-    sm_version: "1.4"
+    python: ["py39"]
+    sm_version: "1.0"
```
Comment (qiyhe) [inline new_images.yml:7]: We are downgrading sagemaker version, is this expected?

Comment (guoqiao1992) [reply new_images.yml:7]: yeah so 1.4 and 1.0 are only part of the tag, because we are releasing 3.1-cpu-py39 for the first time, I set sm_version to 1.0

Comment (qiyhe) [reply new_images.yml:7]: I see, that make sense

File: smsparkbuild/py39/pyproject.toml (hunk 7 above)
Comment (qiyhe) [reply smsparkbuild/py39/pyproject.toml:3]: Did not see it get changed, I assume there will be another commit fixing it

File: test/resources/code/scala/hello-scala-spark/hello-scala-spark.sbt (hunk 5 above)
Comment (guoqiao1992) [reply test/resources/code/scala/hello-scala-spark/hello-scala-spark.sbt]: changed, waiting for test code build job to pass

**Review by qiyhe — APPROVED** [review]
Build succeeded, commit looks good

**Pull request merged**